hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/SqlPageCodec.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl.protocol.codec.builtin;
+
+import com.hazelcast.client.impl.protocol.ClientMessage;
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.sql.SqlColumnType;
+import com.hazelcast.sql.impl.client.SqlPage;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hazelcast.client.impl.protocol.ClientMessage.BEGIN_FRAME;
+import static com.hazelcast.client.impl.protocol.ClientMessage.END_FRAME;
+import static com.hazelcast.client.impl.protocol.codec.builtin.CodecUtil.fastForwardToEndFrame;
+
+public final class SqlPageCodec {
+
+    private SqlPageCodec() {
+    }
+
+    @SuppressWarnings({"unchecked", "checkstyle:CyclomaticComplexity"})
+    public static void encode(ClientMessage clientMessage, SqlPage sqlPage) {
+        clientMessage.add(BEGIN_FRAME.copy());
+
+        // Write the "last" flag.
+        byte[] content = new byte[]{(byte) (sqlPage.isLast() ? 1 : 0)};
```

Comment:
Although it may complicate the implementation a little bit, we can piggyback this into the `BEGIN_FRAME` to get rid of the overhead of an extra frame.

In fact, we can also include the `columnTypeIds` in the `BEGIN_FRAME` to get rid of one more frame overhead. That would require some changes in the `ListIntegerCodec` implementation or we can do it in the loop that we generate `columnTypeIds`

If such an overhead is not an issue, these can stay as it is.

Reply:
After thinking about this again, if we want to also include the `columnTypeIds` in the `BEGIN_FRAME`, we had to write the length of the byte array of the encoded list of the integers because, in the future, we might want to extend this codec with new fix sized parameters and to determine newly added ones we need to know where the byte array of `columnTypeIds` end. This will also be different than what we have in the other codecs (encoding a variable-sized list into the initial frame), so we might just want to include `isLast` bit in the `BEGIN_FRAME`.

But I am fine with making this codec an exception to get rid of an extra frame overhead for `columnTypeIds`. So, including or not including `columnTypeIds` in the `BEGIN_FRAME`, is fine for me

Reply:
My gut feeling is that in many cases the overhead on this additional frame should be minimal comparing to the total serialization time. I can go on with both suggested approaches but have a slight preference towards the simpler implementation (current).